### PR TITLE
Skip Task resource accounting for Fargate 1.3.0 launch type

### PR DESCRIPTION
**Pull request by prateekchaudhry**
### Summary

ECS backend resource accounting can be different for Fargate (1.3.0) than for ECS/EC2 or ECS-Anywhere. This can cause issues with task resource accounting in ECS Agent, if it is not in sync with backend accounting. This behavior can be turned off for Fargate tasks and all resource availability can be delegated to Fargate backend instead.

### Implementation details

* ACS Payload has field `LaunchType` which takes value `FARGATE` for Fargate tasks. For such tasks, do not queue them up and launch them up straightaway. Their resources never get accounted in HostResourceManager.
* HostResourceManager release resources is not a problem as it is an idempotent release. If a task does not exist in HostResourceManager's accounting, this operation will be a no-op.

### Testing

* Added Integ tests to validate HostResourceManager still consumes resources for non fargate tasks, and does not consume resources for fargate tasks
* Manual testing
Both `EC2 and External` -
	- Queued 1 task after the other, enough memory available for 2nd task
		- 2nd task starts without waiting for 1st task to stopped

	- Queued 1 task, Stopped it with stop timeout of 30s, started new task with conflicted resource (CPU)
		- 2nd task starts after 1st task goes to stopped

New tests cover the changes: 
Yes
### Description for the changelog

Skip Task resource accounting for Fargate 1.3.0 launch type
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/engine_unix_integ_test.go [hunk 1]
```diff
@@ -1381,3 +1381,98 @@ func TestHostResourceManagerStopTaskNotBlockWaitingTasks(t *testing.T) {
...
+			Name:       "TestHostResourceManagerRandomLaunchTypeBehavior",
+			LaunchType: "RaNdOmStrInG",
+		},
+		{
+			Name:       "TestHostResourceManagerEmptyLaunchTypeBehavior",
+			LaunchType: "",
+		},
+	}
+
+	for _, tc := range testCases {
```
Comment (amogh09) [inline agent/engine/engine_unix_integ_test.go:1413]: How about running each test case as a subtest with `t.Run`?

File: agent/engine/engine_unix_integ_test.go [hunk 2]
```diff
@@ -1381,3 +1381,98 @@ func TestHostResourceManagerStopTaskNotBlockWaitingTasks(t *testing.T) {
...
+		testTimeout := 1 * time.Minute
+		taskEngine, done, _ := setupWithDefaultConfig(t)
+		defer done()
+
+		stateChangeEvents := taskEngine.StateChangeEvents()
+
+		taskArn := fmt.Sprintf("IntegTaskArn")
```
Comment (amogh09) [inline agent/engine/engine_unix_integ_test.go:1420]: `Sprintf` is redundant here.

File: agent/engine/engine_unix_integ_test.go [hunk 3]
```diff
@@ -1381,3 +1381,98 @@ func TestHostResourceManagerStopTaskNotBlockWaitingTasks(t *testing.T) {
...
+		testTask := createTestTask(taskArn)
+		testTask.Memory = int64(768)
+		testTask.LaunchType = tc.LaunchType
+
+		// create container
+		A := createTestContainerWithImageAndName(baseImageForOS, fmt.Sprintf("A"))
```
Comment (amogh09) [inline agent/engine/engine_unix_integ_test.go:1426]: `Sprintf` is redundant here. Also can we please use a more descriptive name than `A`?

File: agent/engine/engine_unix_integ_test.go [hunk 4]
```diff
@@ -1381,3 +1381,98 @@ func TestHostResourceManagerStopTaskNotBlockWaitingTasks(t *testing.T) {
...
+		A.EntryPoint = &entryPointForOS
+		A.Command = []string{"trap shortsleep SIGTERM; shortsleep() { sleep 6; exit 1; }; sleep 10"}
+		A.Essential = true
+		A.StopTimeout = uint(6)
+		testTask.Containers = []*apicontainer.Container{
+			A,
+		}
+
+		// Stop task payloads from ACS for the tasks
+		stopTask := createTestTask(fmt.Sprintf("IntegTaskArn"))
```
Comment (amogh09) [inline agent/engine/engine_unix_integ_test.go:1436]: `Sprintf` is redundant.

File: agent/engine/engine_unix_integ_test.go [hunk 5]
```diff
@@ -1381,3 +1381,98 @@ func TestHostResourceManagerStopTaskNotBlockWaitingTasks(t *testing.T) {
...
+
+			// At this time, stopTask is received, and SIGTERM sent to task
+			// but the task is still RUNNING due to trap handler
+			assert.Equal(t, apitaskstatus.TaskRunning, testTask.GetKnownStatus(), "task known status should be RUNNING")
+			assert.Equal(t, apitaskstatus.TaskStopped, testTask.GetDesiredStatus(), "task desired status should be STOPPED")
+			// Verify resources are properly consumed in host resource manager, and not consumed for Fargate
+			if tc.LaunchType == "FARGATE" {
+				assert.False(t, taskEngine.(*DockerTaskEngine).hostResourceManager.checkTaskConsumed(testTask.Arn), "fargate task resources should not be consumed")
+			} else {
+				assert.True(t, taskEngine.(*DockerTaskEngine).hostResourceManager.checkTaskConsumed(testTask.Arn), "non fargate task resources should not be consumed")
```
Comment (amogh09) [inline agent/engine/engine_unix_integ_test.go:1470]: Typo in "non fargate task resources should not be consumed"? Should be "non fargate task resources should be consumed"?

File: agent/engine/task_manager.go [hunk 6]
```diff
@@ -285,8 +285,9 @@ func (mtask *managedTask) waitForHostResources() {
 		return
 	}
 
-	if !mtask.IsInternal && !mtask.engine.hostResourceManager.checkTaskConsumed(mtask.Arn) {
+	if !mtask.IsLaunchTypeFargate() && !mtask.IsInternal && !mtask.engine.hostResourceManager.checkTaskConsumed(mtask.Arn) {
```
Comment (amogh09) [inline agent/engine/task_manager.go:288]: Is this the only change we need? The PR description says "Their resources never get accounted in HostResourceManager.". What prevents a Fargate task's resources from being accounted in HostResourceManager?

File: agent/engine/engine_unix_integ_test.go (hunk 1 above)
Comment (prateekchaudhry) [reply agent/engine/engine_unix_integ_test.go:1413]: Thanks! let me update with a follow up

File: agent/engine/task_manager.go (hunk 6 above)
Comment (prateekchaudhry) [reply agent/engine/task_manager.go:288]: All tasks have a managedTask, and go through [overseeTask](https://github.com/aws/amazon-ecs-agent/blob/08f9dc120c9a23c3e5c7f969e48398a243066819/agent/engine/task_manager.go#L197) on start up. It is a linear path to here on receiving a Start task request on an already started instance. So this should be the only needed change.

This change prevents queuing up task to go through [waitingTaskQueue](https://github.com/aws/amazon-ecs-agent/blob/08f9dc120c9a23c3e5c7f969e48398a243066819/agent/engine/docker_task_engine.go#L392-L413) processing for Fargate tasks, and hence they will never be accounted by HostResourceManager

Comment (amogh09) [reply agent/engine/task_manager.go:288]: Thanks!

**Review by amogh09 — APPROVED** [review]

**Review by yinyic — APPROVED** [review]

**Review by amogh09 — DISMISSED** [review]

**Review by yinyic — APPROVED** [review]

**Review by amogh09 — APPROVED** [review]

**Pull request merged**